### ISSUE TITLE
SARIMAX dynamic prediction

### ISSUE THREAD

**Issue by Demyxter** [statsmodels/statsmodels#4012]
Hey there. Currently i'm trying to fit a SARIMAX model in order to be able to predict incoming sales. However, i'm encountering a problem when trying to use the .predict method, specially with the dynamic argument. From the documentation:

> dynamic : boolean, int, str, or datetime, optional
>
> Integer offset relative to start at which to begin dynamic prediction. Can also be an absolute date string to parse or a datetime type (these are not interpreted as offsets). Prior to this observation, true endogenous values will be used for prediction; starting with this observation and continuing through the end of prediction, forecasted endogenous values will be used instead.

i'm assuming this option enables me to do a rolling prediction, using the predicted sales as AR parameters to predict future ones, right?

```
mod = SARIMAX(to_predict[:TrnSize],predictors[:TrnSize],order =(3,0,3),enforce_invertibility=False)

res  = mod.fit(disp=False)

```

However, when i fit my model, and print my predictions, i get the same predictions, wheter i'm using dynamic True or False. I also tried to use integer values for dynamic, getting the same results.
```

print(res.predict(start=TrnSize,end=TrnSize+TstSize-1,exog=predictors[TrnSize:TrnSize+TstSize],dynamic=True))

2015-05-31    1638.219412
2015-06-30    1838.222821
2015-07-31    1897.165998
2015-08-31    1827.262010
2015-09-30    1873.872274
2015-10-31    1655.454894
2015-11-30    1522.591954
2015-12-31    1170.836379
2016-01-31    1130.656283
2016-02-29    1364.341904
```

```
print(res.predict(start=TrnSize,end=TrnSize+TstSize-1,exog=predictors[TrnSize:TrnSize+TstSize],dynamic=False))

2015-05-31    1638.219412
2015-06-30    1838.222821
2015-07-31    1897.165998
2015-08-31    1827.262010
2015-09-30    1873.872274
2015-10-31    1655.454894
2015-11-30    1522.591954
2015-12-31    1170.836379
2016-01-31    1130.656283
2016-02-29    1364.341904

```

From my understanding, using dynamic True/False should yield different results since i'm using AR parameters, right? what am i doing wrong here?
Currently using version 0.8.0.

**Comment by josef-pkt** [comment]
From your predict dates it looks like you do out-of-sample prediction (which is always out-of-training sample). In this case there is only one `dynamic` for `forecasting`. (I never remember the direction for true or false in dynamic). So the result is expected.

However, I don't know what's the easiest workaround to filter the additional test data to get a one step ahead prediction.

**Comment by Demyxter** [comment]
Yes, i indeed do out-of-sample prediction. So i won't be able to do the "rolling prediction", then? i tried to do one step ahead predictions,  add the predicted value to the training sample and re-fitting the model, but this process is painfully slow.

**Comment by josef-pkt** [comment]
We had some issues on this, the only closely related issues that I find right now is #3623, and maybe #2577

@ChadFulton knows better, 

Aside: If you have to refit, then don't forget to use start_params with the previous estimate to speed up estimation. I assume this works also for the statespace models.

**Comment by ChadFulton** [comment]
I think the model is already doing what you wanted (what you called "rolling prediction"). I'm curious what you thought the default setting was doing?

In the case of state space models, the `dynamic` argument is actually pretty straightforward: it just tells the model when to cut off the observed data for the purpose of forecasting.

Consider the three following specifications

```python
mod1 = sm.tsa.SARIMAX(endog[:5], exog=exog[:5], order=(1, 0, 0))
mod2 = sm.tsa.SARIMAX(endog[:10], exog=exog[:10], order=(1, 0, 0))
mod3 = sm.tsa.SARIMAX(np.r_[endog[:5], [np.nan]*5], exog=exog[:10], order=(1, 0, 0))
```

The first model only includes the first 5 datapoints, the second model includes the first 10 datapoints, and the third replaces `endog` values from 5-10 with np.nan (i.e. missing data).

Now, suppose that we evaluate each of the models using the same parameters and compare the following:

- For `mod1`, do out-of-sample forecasting for observations 5-10
- For `mod2`, do in-sample prediction with dynamic prediction starting at observation 5
- For `mod3`, do in-sample prediction with `dynamic=False` (the default)

```python
params = [1, 0.5, 1]

res1 = mod1.smooth(params)
res2 = mod2.smooth(params)
res3 = mod3.smooth(params)

print(res1.predict(start=0, end=9, exog=exog[5:10, None]))
print(res2.predict(start=0, end=9, dynamic=5))
print(res3.predict(start=0, end=9))
```

These predict commands will all return the same results, because they are only using data from observations 0-4 for the purposes of prediction.

**Comment by Demyxter** [comment]
oooh, wow, thanks. From reading the documentation: 

>  Prior to this observation, true endogenous values will be used for prediction; starting with this observation and continuing through the end of prediction, forecasted endogenous values will be used instead.

Since i was doing out-of-sample predictions, there would be no true endogenous values. So i just assumed that with `dynamic = False` ,the model would just assume them to be 0 and forecast based only on the exogenous values. Guess my english understanding still have quite some way to go.

Anyway, thanks for the help (again).

**Comment by ChadFulton** [comment]
The documentation should be improved, when I originally wrote it I didn't fully understand what was going on in `dynamic`.

(and actually I think our implementation in `KalmanFilter._predic` is more complicated than it needs to be)

> Prior to this observation, true endogenous values will be used for prediction; starting with this observation and continuing through the end of prediction, forecasted endogenous values will be used instead.

This is a technically true statement, but what's missing from this is that using the forecasted endogenous values is exactly the same as having missing data.

Technically, the updating step of the Kalman filter usually is:

a_{t|t} = a_t + K v_t

where v_t is the forecast error, equal to endog - forecast. But if we substitute forecasted values for endog, then the forecast error is equal to forecast - forecast = 0, always. This means the updating step is trivial:

a_{t|t} = a_t

And this is exactly what happens when there is missing data.

**Comment by gdalle** [comment]
Hi ! Thanks for the answers :)
So from what I gather, the dynamic prediction of mod2 goes like : 
- I know endog[0], let's guess endog[1]
- I know endog[0] and I guessed endog[1] wrong but I know the real value know, so let's guess endog[2] from it
- ...
- Same thing for endog[4]
- From endog[5] onwards the model cannot correct its own guesses with the true values and so it keeps on building predictions on more and more fragile grounds
Am I correct ? If so, is it possible to let the model predict more than one step ahead between indices 0 and 4 ?

**Comment by ehsan2167** [comment]
Dear ChadFulton
I am Ehsan, and I am working on my project which I have to forecast in future (120 steps), but my result is not accurate, do you have any suggestion for me?
Thanks in advance
Ehsan    

![captwwure](https://user-images.githubusercontent.com/17020971/39026824-a521f4ac-4414-11e8-9baf-3bad087112a1.PNG)

**Comment by ChadFulton** [comment]
What is your concern specifically? The forecasts in the graphic you attached look pretty reasonable to me. I don't know how much better you'll be able to do with a one-variable model, and ten years ahead is a really long time. My experience is that forecast quality usually substantially degrades after even a few months.

**Comment by ehsan2167** [comment]
Thanks, we will be Intouch.Best regards Ehsan
    On Friday, April 20, 2018, 8:26:19 PM CDT, ChadFulton <notifications@github.com> wrote:  
 
 
What is your concern specifically? The forecasts in the graphic you attached look pretty reasonable to me. I don't know how much better you'll be able to do with a one-variable model, and ten years ahead is a really long time. My experience is that forecast quality usually substantially degrades after even a few months.

**Comment by bashtage** [comment]
Closing as answered.

**Comment by GithubBiswajit** [comment]
I am trying to build a Time series model using Sarimax, My data has a Seasonality component, that every 7 days there is a peak. 

model = SARIMAX(df1['Sales'],order=(4,1,0),seasonal_order=(0, 0, 0, 7))
results = model.fit()
fcast = results.predict(len(df1),len(df1)+7,typ='levels').rename('SARIMA Forecast')

The output of the prediction does not show that it is showing a flat line around mean value

![Untitled](https://user-images.githubusercontent.com/67324094/85406008-cbec7700-b57e-11ea-94d4-28f2f0a2e2e9.png)

**Comment by ChadFulton** [comment]
With `seasonal_order=(0, 0, 0, 7)`, you aren't specifying any seasonal dynamics, so the model is not actually a seasonal one.

**Comment by GithubBiswajit** [comment]
> With `seasonal_order=(0, 0, 0, 7)`, you aren't specifying any seasonal dynamics, so the model is not actually a seasonal one.

Thanks, Mr. Fulton, 

Even without that also ai am not getting adequate results, it is predicting mean value only.
when I have run 
auto_arima(df1['Sales'],seasonal=True,m=7).summary()
it is suggesting SARIMAX(4, 1, 0).. without any seasonal dynamics , however there is a clear cut seasonality component 
how to handle this 

model = SARIMAX(train['Sales'],order=(4,1,0))
results = model.fit()
results.summary()

start=len(train)
end=len(train)+len(test)-1
predictions = results.predict(start=start, end=end, dynamic=False, typ='levels').rename('SARIMA Predictions')


model = SARIMAX(df1['Sales'],order=(4,1,0))

results = model.fit()
fcast = results.predict(len(df1),len(df1)+7,typ='levels').rename('SARIMA Forecast')

# Plot predictions against known values
title = 'Sales Data'
ylabel='Lacs'
xlabel=''

ax = test['Sales'].plot(legend=True,figsize=(12,6),title=title)
fcast.plot(legend=True)
ax.autoscale(axis='x',tight=True)
ax.set(xlabel=xlabel, ylabel=ylabel);

please advise accordingly

**Comment by bashtage** [comment]
Could you edit your post and surround code with triple back-ticks \`\`\`.

Set the seasonal AR or MA to be > 0. So something like (1,0,0,7) or (0,0,1,7).
